### ISSUE TITLE
Hybrid Android app – Copilot – Unable to switch to own account from copilot account

### ISSUE THREAD

**Issue by IuliiaHerets** [Expensify/App#51042]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** v9.0.50-0
**Reproducible in staging?:** N
**Reproducible in production?:** N
**If this was caught on HybridApp, is this reproducible on New Expensify Standalone?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:** https://expensify.testrail.io/index.php?/tests/view/5097100
**Email or phone of affected tester (no customers):** applausetester+jp_2171024@applause.expensifail.com
**Issue reported by:** Applause Internal Team

## Action Performed:
1. Go to Hybrid app
2. Login as the account that has full copilot access
3. Tap on Settings/header arrows
4. Switch to copilot accounts
5. Tap on Settings/header arrows again
6. Switch to your own accounts

## Expected Result:
User able to switch to own account from copilot account

## Actual Result:
Error appears: `Oops something went wrong`

## Workaround:
Unknown

## Platforms:

- [ ] Android: Standalone
- [x] Android: HybridApp
- [ ] Android: mWeb Chrome
- [ ] iOS: Standalone
- [ ] iOS: HybridApp
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/05af81de-e678-484e-b33e-b7ebdb662fc7

 
</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Issue Owner</summary>Current Issue Owner: @dangrous</details>

**Comment by IuliiaHerets** [comment]
Sorry, was confused with blocker label

**Comment by jasperhuangg** [comment]
@IuliiaHerets next time add both labels, they're used to indicate that the bug could either be from the front-end or the back-end.

**Comment by jasperhuangg** [comment]
[These](https://www.expensify.com/_devportal/tools/logSearch/#sort=asc&size=10000&query=blob%3A%22401%20Not%20authenticated%20to%20this%20partner%22%20AND%20email%3A%22applausetester%2Bjp_1171024%40applause.expensifail.com%22%20AND%20timestamp%3A%5B2024-10-17T11%3A22%3A48.058Z%20TO%202024-10-17T13%3A22%3A48.058Z%5D) logs seem suspect, I'll try to reproduce on my own account to see if I run into the same issue. Need to wait for the beta cache to refresh though.

**Comment by jasperhuangg** [comment]
Strange, checking CreateLogin calls in the logs, it seems Android is the only platform that sends partnerID=16 (which is the Android partnerID). All other platforms (even iPhone), send partnerID=83 (which is the NewDot partnerID).

**Comment by jasperhuangg** [comment]
Going to ask in Slack for some more context about how we decide which partnerID to send with each request since I lack context.

**Comment by jasperhuangg** [comment]
Ah I was finally able to repro. The tester is testing using HybridApp. When we sign into HybridApp on Android/iPhone and then switch to NewDot, the authToken is created with partnerID = 14 or 16 (i.e. PARTNER_NAME_IPHONE or PARTNER_NAME_ANDROID). When we try to call CreateLogin with an authToken with those partnerIDs, it isn't recognized [here](https://github.com/Expensify/Auth/blob/bd180f230c89ad5765b8803626d66faefcd5b55b/auth/command/CreateLogin.cpp#L54-L64) and throws.

We need to either:
- Add PARTNER_NAME_IPHONE and PARTNER_NAME_ANDROID to the list [here](https://github.com/Expensify/Auth/blob/bd180f230c89ad5765b8803626d66faefcd5b55b/auth/command/CreateLogin.cpp#L54-L64).
- Ensure that when someone taps "Try New Expensify" on mobile, we create the authToken with partnerID = 83 (i.e. PARTNER_NAME_NEWDOT) since that's the behavior for the New Expensify App.

Gonna ask in Slack for some feedback on these options.

**Comment by jasperhuangg** [comment]
NewDot Copilot lives behind a beta, so I don't think we need to block deploy on this.

**Comment by jasperhuangg** [comment]
#engineering chat post asking for feedback [here](https://expensify.slack.com/archives/C03TQ48KC/p1729192777695799)

**Comment by dylanexpensify** [comment]
Hey @twisterdotcom! I'm heading out on parental leave so reassigning this! TY! 🙇‍♂️

**Comment by jasperhuangg** [comment]
Have some feedback in the Slack thread to look into calling `SignInWithShortLivedAuthToken` to get a NewDot token when switching to NewDot. It seems that's what we do in other cases when switching from OldDot to NewDot.

**Comment by jasperhuangg** [comment]
Have my hands full working on a design doc for a high priority #migrate initiative, probably can get to this start of next week.

**Comment by jasperhuangg** [comment]
Same!

**Comment by twisterdotcom** [comment]
How is this going @jasperhuangg?

**Comment by jasperhuangg** [comment]
Haven't been able to prioritize this due to a higher priority #migrate project that I'm still working on, thanks for the bump.

The latest on the solution I'm going to be implementing is in [this](https://expensify.slack.com/archives/C03TQ48KC/p1729271040340379) Slack thread.

**Comment by jasperhuangg** [comment]
Was able to find some time to revisit this today. I'm asking some more questions [here](https://expensify.slack.com/archives/C03TQ48KC/p1732044755052569) since I'm pretty unfamiliar with how our HybridApp code works.

**Comment by jasperhuangg** [comment]
@dangrous I find that I keep running into hiccups in discussions because I wasn't the one who implemented delegated access in NewDot, is there any chance you could take this off my hands?

**Comment by twisterdotcom** [comment]
Which project do we think this fits in? Where did we handle Copilot before?

**Comment by jasperhuangg** [comment]
[Here's](https://docs.google.com/document/d/18G6J5bl61VnVa5CoKpRXhM0MhkXDwpm87GVitZrSCVg/edit?tab=t.0) the doc, looks like we bunched it in with #wave-control before. So maybe #expense?

**Comment by dangrous** [comment]
yeah i think it was #wave-control and we were handling the remaining bugs in #quality?

I can take this over but I may keep bothering you with questions (see Slack haha) because I am very bad at token stuff. But I can try!

**Comment by dangrous** [comment]
nearly there! App PR is approved by C+ so just getting the internal reviewer's approval

**Comment by dangrous** [comment]
App is merged, waiting on deploy and then backend review, should be able to test this shortly

**Comment by dangrous** [comment]
okay, both bits of the solution for this have been merged! We should be able to test this out and close it out - eta tomorrow.

**Comment by twisterdotcom** [comment]
Nobody needs paying here right?

**Comment by dangrous** [comment]
I think @mollfpr will need payment for C+ review.

But I still need to double check that this is working on prod; @war-in suggests that it might not be yet. I'm about to try locally and on browserstack

**Comment by dangrous** [comment]
Okay so this doesn't work for me on hybrid app (though it works on regular android app). I don't know what changed!

**Comment by war-in** [comment]
I'm gonna build the app from the main branches and try to find some logs. I'll post them here 🫡

**Comment by war-in** [comment]
I managed to make it work by merging the main branches (not sure what exactly changed but it works 🤷).
Now I have a question about [those](https://github.com/Expensify/App/blob/bc956d0e392b4c670337b7c6277ea364c7515749/src/libs/actions/Delegate.ts#L206) few set actions

A few lines above we update the `authToken` and `encryptedAuthToken` but immediately we override them by calling the set with old (stashed) data.
I think we should update tokens after calling `set`. Do you think it makes sense?

**Comment by dangrous** [comment]
oh huh, yeah I think you're right!

**Comment by dangrous** [comment]
Hi Melvin, still in progress. Aiming for early next week

**Comment by dangrous** [comment]
Estimated merge is on Friday!

**Comment by dangrous** [comment]
50% there - ND PR is merged, and the Mobile-Expensify one is up for final review from SWM

**Comment by dangrous** [comment]
Oh wait maybe we don't get comments on deploys in M-E? Checking in Slack, but hopefully that means we can get this re-tested and closed!

**Comment by Julesssss** [comment]
Yeah, there's no staging branch or deploy on Mobile-E. Once merged to main the next App build (including CP builds) will include the mobile changes.

**Comment by dangrous** [comment]
ah great! @IuliiaHerets are you able to retest this?

**Comment by IuliiaHerets** [comment]
@dangrous Issue is not reproduced now 

https://github.com/user-attachments/assets/4acedeb3-8594-4494-8c6b-9154707c83c2

**Comment by dangrous** [comment]
Amazing! I think we can close this - I don't believe any payment is required. If someone disagrees, ping here!
